Improve Feature detection

sonicaj opened the pull request:
This PR aims to improve how py-libzfs handles features along differing kernel versions. The intention is to use autoconf to handle conflicts regarding versions and simplify python code

sonicaj committed:
Improve feature detection

This commit adds support for building py-libzfs with autoconf. Till now tests for header file 'zfs.h' have been completed.
Ticket: #47268

sonicaj committed:
libzfs.h header file usages cleared

sonicaj committed:
Minor improvements

sonicaj committed:
Usages of libzfs_core.h conflicts completed

sonicaj committed:
More conflicts resolved

sonicaj committed:
Code formatted

sonicaj committed:
Generate a pxi file for Cython

igalic commented on configure.ac:
```diff
@@ -0,0 +1,246 @@
+-I/usr/src/sys/cddl/contrib/opensolaris/uts/common/fs/zfs/ -I/usr/src/sys/cddl/compat/opensolaris/sys/ -I/usr/src/sys/cddl/compat/opensolaris/ -I/usr/src/cddl/lib/libumem 
+-I/usr/src/usr/src/cddl/compat/opensolaris/ -I/usr/src/usr/src/cddl/compat/opensolaris -I/usr/src/cddl/compat/opensolaris/lib/libumem -I/usr/src/cddl/contrib/opensolaris/lib/libzpool/common 
+-I/usr/src/sys/cddl/contrib/opensolaris/common/zfs -I/usr/src/sys/cddl/contrib/opensolaris/uts/common/sys -I/usr/src/cddl/contrib/opensolaris/head -I/usr/src/cddl/contrib/opensolaris/lib/libnvpair 
+-I/usr/src/cddl/contrib/opensolaris/lib/libuutil/common -I/usr/src/cddl/contrib/opensolaris/lib/libzfs/common -I/usr/src/cddl/contrib/opensolaris/lib/libzfs_core/common ${CFLAGS} -I/usr/include"
```
maybe make `/usr/src` an overrideable variable. some people like to checkout their source tree in esoteric places, while others may have many different versions checked out

igalic commented on Makefile.in:
```diff
@@ -0,0 +1,11 @@
+PYTHON ?= /usr/local/bin/python2.7
```
isn't 3.6 the default these days?

gronke replied:
There's even 3.7 coming with the latest packages. I find it okay to leave the current default (Python 2.7) as long as it can be overridden with an env variable.

sonicaj replied:
It can be passed from the command line `make install PYTHON=/usr/local/bin/python`

igalic reviewed (commented):
I love this!

gronke commented on .gitignore:
```diff
@@ -1,4 +1,9 @@
-venv/
+autom4te.cache/
+.idea/
```
Why is this in the .gitignore? Shouldn't that be in a users global .gitignore?

gronke commented on configure:
```diff
@@ -0,0 +1,5018 @@
+# Printing a long string crashes Solaris 7 /usr/bin/printf.
+as_echo='\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\'
+as_echo=$as_echo$as_echo$as_echo$as_echo$as_echo
+as_echo=$as_echo$as_echo$as_echo$as_echo$as_echo$as_echo
```
Can you please explain what this is good for? Why exactly this many backslashes?

sonicaj replied:
Configure file is dynamically generated by autoconf - this is probably being done to support cross platform runs ?

gronke replied:
I agree with @igalic that this file should be excluded from the git repository.

william-gr replied:
There is no harm in having it here.

gronke replied:
The file is too complex to reasonably review it. The work overhead of keeping it audited is pretty  high, so that there is the risk of malicious code within the file. For security reasons I recommend to drop the file from the source repository.

william-gr replied:
Thank you for your opinion but we are going to keep it. Security is not a concern for us.
If cpython and dozens of other projects do it, I dont see why we should do differently for these pointed out reasons.

gronke replied:
> Security is not a concern for us.

For some of us it is. At the current time [libiocage](https://github.com/iocage/libiocage/) relies on this module with elevated privileges. I would prefer py-libzfs to contain no code that the authors did not understand. And in case of a auto-generated file, I don't see the reason to raise the overhead of understanding each and every bit of it versus letting it become generated by autoconf from files that are more human-readable.

> If cpython and dozens of other projects do it, I dont see why we should do differently for these pointed out reasons.

IMO they are doing it wrong or have pretty aware reviewers.

william-gr replied:
You are not required to use it, you can generate it yourself if you dont want to audit configure.

igalic replied:
python's configure.ac is probably the cleanest m4 code i've ever seen in my life, and still, it's configure script is 18k lines. but, what they somehow manage is this:

```
 ~/s/p/cpython  ⎇  rm configure
 ~/s/p/cpython  ⎇ *  autoreconf -if
 ~/s/p/cpython  ⎇ 
 ~/s/p/cpython  ⎇  git diff
 ~/s/p/cpython  ⎇  # nothing to see here
```

gronke replied:
> You are not required to use it, you can generate it yourself if you dont want to audit configure.

That works for me! I'd love to see it being used by the ports package:

```
USES=		python
```
can be changed to
```
USES=		autoreconf python
```

gronke commented on configure:
```diff
@@ -0,0 +1,5018 @@
+ac_ext=c
+ac_cpp='$CPP $CPPFLAGS'
+ac_compile='$CC -c $CFLAGS $CPPFLAGS conftest.$ac_ext >&5'
+ac_link='$CC
```
This variable is defined further above.

sonicaj replied:
dynamic generation of file by autoconf

gronke commented on configure:
```diff
@@ -0,0 +1,5018 @@
+# there to prevent editors from complaining about space-tab.
+# (If _AS_PATH_WALK were called with IFS unset, it would disable word
+# splitting by setting IFS to empty value.)
+IFS=" ""	$as_nl"
```
Isn't it easier to define a variable called `$as_tab` and omit the comment?

sonicaj replied:
dynamic generation of file by autoconf

gronke commented on configure:
```diff
@@ -0,0 +1,5018 @@
+ac_ext=c
+ac_cpp='$CPP $CPPFLAGS'
+ac_compile='$CC -c $CFLAGS $CPPFLAGS conftest.$ac_ext >&5'
+ac_link='$CC
```
What does the `#(((((` pattern do? Never saw this before and it confuses me 😅

gronke commented on configure.ac:
```diff
@@ -0,0 +1,246 @@
+                [#include <sys/fs/zfs.h>],
+                [
+                        int i = sizeof(pool_scrub_cmd_t);
+			i = i; /* ; */
```
This line uses tabs while the line directly above indents with spaces.

gronke reviewed (changes requested):
I've added a few questions, but did not find the time to do a full audit of the `configure` file yet.

sonicaj committed:
Reviews Addressed(1)

Following fixes are in this commit:
1) Allows users to pass a variable for freebsd_src tree
2) Fixes gitignore issue
3) Indent issue fixed

sonicaj committed:
Move pxi file to pxd folder and remove unused files

sonicaj committed:
Move pxi file to pxd folder and remove unused m4 macro

sonicaj committed:
Check for a member in sendflags_t struct

sonicaj committed:
Use config.pxi in pxd files

sonicaj committed:
Check LZC_SEND_FLAG_EMBED_DATA enum presence

sonicaj committed:
Remove version conditionals in libzfs.pyx

sonicaj committed:
Todo completed

sonicaj committed:
All version conditionals removed

sonicaj committed:
Remove trueos usage in conditionals

sonicaj committed:
Remove experimental usages from conditionals

sonicaj committed:
Tidying up code

sonicaj committed:
Removed all usages of freebsd_version, trueos and experimental

sonicaj committed:
Tidying configure.ac

sonicaj committed:
update readme

igalic commented on configure.ac:
```diff
@@ -0,0 +1,315 @@
+	#include <sys/types.h>
+	#include <sys/mnttab.h>
+	#include <sys/fs/zfs.h>
+	#include <sys/zfs_ioctl.h>
```
can you give us a quick run-down on which features the `sys/zfs_ioctl` header is used for?

sonicaj replied:
When checking for `libzfs.h` header, `sys/zfs_ioctl` is a header which is being used by `libzfs.h` and thus we include it as a dependency.
`#include <sys/zfs_ioctl.h>` Taken from `libzfs.h`

igalic replied:
aye! thanks for the explanation!

sonicaj replied:
No problem

igalic reviewed (commented):
❓
